src/is_subscribed.rs

```diff
@@ -0,0 +1,107 @@
+/// find if the current user is subscribed to the ?topic
+///
+/// if topic is None, then check if the user is subscribed to any topic. This is indicated by
+/// `subscribed: true` with `topics: []` in the user's data.
+///
+/// the user is fetched from the session cookie. If the user is not logged in, the tracker cookie
+/// is used in its place.
+#[ft_sdk::data]
+fn is_subscribed(
+    ft_sdk::Query(topic): ft_sdk::Query<"topic", Option<String>>,
+    sid: ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
+    tid: ft_sdk::Cookie<"fastn-tid">,
+    mut conn: ft_sdk::Connection,
+) -> ft_sdk::data::Result {
+    let data = match user_data_from_sid_or_tid(sid, tid, &mut conn) {
+        Ok(d) => d,
+        Err(_) => return ft_sdk::data::json(serde_json::json!({ "subscribed": false })),
+    };
+
+    let subscribed = check_if_subscribed(topic, data);
+
+    ft_sdk::data::json(serde_json::json!({ "subscribed": subscribed }))
+}
+
+fn check_if_subscribed(topic: Option<String>, user_data: serde_json::Value) -> bool {
+    let subscribed = if let Some(topic) = topic {
+        user_data
+            .as_object()
+            .map(|v| v.get("subscription"))
+            .flatten()
+            .map(|v| v.as_object())
+            .flatten()
+            .map(|v| v.get("topics"))
+            .flatten()
+            .map(|v| v.as_array())
+            .flatten()
+            .map(|v| v.iter().any(|t| t.as_str() == Some(&topic)))
+            .unwrap_or(false)
+    } else {
+        user_data
+            .as_object()
+            .map(|v| v.get("subscription"))
+            .flatten()
+            .map(|v| v.as_object())
+            .flatten()
+            .map(|v| v.get("subscribed"))
+            .flatten()
+            .map(|v| v.as_bool())
+            .flatten()
+            .unwrap_or(false)
+    };
+
+    subscribed
+}
+
+fn user_data_from_sid_or_tid(
+    ft_sdk::Cookie(user_id_from_session): ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
+    ft_sdk::Cookie(user_id_from_tracker): ft_sdk::Cookie<"fastn-tid">,
+    conn: &mut ft_sdk::Connection,
+) -> Result<serde_json::Value, ft_sdk::Error> {
```

Comment:
```suggestion
) -> Result<ft_sdk::AppData, ft_sdk::Error> {
```

ft_sdk::AppData somewhat similar to ft_sdk::auth::ProviderData

Reply:
not needed right now. An important for the future ;)